### DISCUSSION TITLE
PersistenceContext EntityEntries Delete and Flush

### DISCUSSION THREAD

**Discussion by danspam** [nhibernate/nhibernate-core#3245] [Q&A] [answered]
We are using the `EntityEntries` in the `PersistenceContext` to inspect all the entities used in a unit of work just before the transaction completes. After some investigation, I have discovered when you delete an entity (which adds a deleted status entry into `EntityEntries`) and then cause a session flush (which can happen automatically), the `EntityEntry` for the deleted entity is removed from the `PersistenceContext`.

Can someone tell me if that is the expected behaviour for deleted entities in `EntityEntries`? Given that there is also a Gone status for the EntityEntry, I was expecting it to remain in the `EntityEntries` but with status changed to Gone. 

If this is the expected behaviour, then it seems the only way to get a complete list of all entities used in a unit of work would be to track them all manually using the event system, or am I missing something?

**Comment by fredericDelaporte** [comment, marked as answer]
Yes, a gone entity is removed from the persistence context. See `EntityDeleteAction`.
https://github.com/nhibernate/nhibernate-core/blob/07c367fb859086970737c35f24fe66c089a034e2/src/NHibernate/Action/EntityDeleteAction.cs#L77-L84

`EntityEntry.PostDelete` is the only thing that does affect the `Gone` status, and that is the only place where it is called. As you can see, it is called after removing the entry from the context.

The `Gone` status is useful for NHibernate internal processing because, during flushes by example, it works on snapshots of the persistence context which may still contain already removed entities.

> If this is the expected behaviour, then it seems the only way to get a complete list of all entities used in a unit of work would be to track them all manually using the event system, or am I missing something?

You could use the persistence context for most entities and track deleted ones through events.

**Reply by danspam** [reply]
Understood. Thanks for the clarification.
